Gas Estimation examples for submit solution

bh2smith opened the pull request:
## Test Plan

```
truffle test test/stableX_large.js -g 
```

 New estimates based on #266 and discussion in #257 


## Cost for  Submitting a Competing Solution

20 GWei with ETH @ ~183 USD

| num Orders   | Gas Used (Million) | USD Cost |
| ------------- |:-------------:| -----:|
| 10 | 2.0   | $7.24 |
| 20 | 3.8 |  $13.76 |
| 24 | 4.6 |  $16.65 |
| 30 | 5.6 |  $20.27 |
| 40 | gas limit exceeded | XXX |

bh2smith committed:
First pass at evaluating utility and disregarded utility on solution submission.

bh2smith committed:
Compiling version of first pass

bh2smith committed:
Updated Unit tests to new objective function.

bh2smith committed:
remove Debug Events.

bh2smith committed:
evaluateUtility can be restricted to view

bh2smith committed:
typos, compile warnings and comment spacing

bh2smith committed:
include brackets in describe blocks because, for example, there are 46 occurences of "addToken" but only one of "addToken()"

bh2smith committed:
Merge branch 'cosmetics' into disregarded_utility

# Conflicts:
#	contracts/StablecoinConverter.sol
#	test/stablecoin_converter.js

bh2smith committed:
Update formulas in smart contract and a couple other suggestions (not passing yet)

bh2smith committed:
revert disregarded utility evaluation chages, because it wasn't working.
Signed-off-by: Ben Smith <bh2smith@gmail.com>

bh2smith committed:
lint

bh2smith committed:
remove Debug events

bh2smith committed:
introduce gas estimation examples for submit solution.

bh2smith committed:
Include more tests to support the claims.

bh2smith committed:
make relevant functions view and pure

josojo commented:
Cool. It is very important to know these numbers. very useful test.

Three small comments:
1. The numbers would probably deviate a lot, if not for all trades the same owners are used, as then new storage would be required, which is more expensive than updating existing storage on-chain.
2. We could avoid some code dublications with my other scripts by separating some functions and definitions into their own .js file
3. while these tests are useful for measuring our exchange, I would prefer to not have them as part of the usual test suite, as testing takes already quite some time..

bh2smith commented:
> The numbers would probably deviate a lot, if not for all trades the same owners are used, as then new storage would be required, which is more expensive than updating existing storage on-chain.

Can you point me specifically at a point in the contract that would be more costly if we had different owners on each touch order?


> We could avoid some code duplications with my other scripts by separating some functions and definitions into their own .js file

I believe I already avoided a lot of code duplication with the `beforeEach` statement in the test file. However, the Basic Trade example found in this file is different than the one used in the main file.

> while these tests are useful for measuring our exchange, I would prefer to not have them as part of the usual test suite, as testing takes already quite some time..

Yes, these tests were not meant to be used in the main code. They are just here as an experiment for gas estimations. This is also why this branch is separate from the other (https://github.com/gnosis/dex-contracts/pull/229) which contains objective function evaluation and should eventually be merged.

bh2smith commented:
@josojo, after a discussion with @koeppelmann today, these gas estimations suggest that we should be putting a limit (of, say 25) on the number of touched orders because if we did not have this, the first submitter would have the serious advantage that overwriting a solution with more that 40 touched orders would exceed block gas limit. I will include this constraint in #229 along with a relevant unit test.

josojo commented:
>The numbers would probably deviate a lot, if not for all trades the same owners are used, as then new storage would be required, which is more expensive than updating existing storage on-chain.

>Can you point me specifically at a point in the contract that would be more costly if we had different owners on each touch order?

In the function addBalance and subtractBalance, it makes a difference, whether the user has already a positive balance, or not. In case he has non, the opcode: Sstore is required to store their balance and hence it is much more expensive then just updating their balance. 

If we settle 40 orders , it is 15000*40 = 1/2 million more gas. Okay, it is not a lot, but probably significant. :) 

>Yes, these tests were not meant to be used in the main code. They are just here as an experiment for gas estimations. This is also why this branch is separate from the other (#229) which contains objective function evaluation and should eventually be merged.

Maybe it would be good to have the part of the main code for easier maintenance, but not being run in each test instance. But I don't know how to implement it, so I am okay with just having it on this separate branch.

josojo commented on test/stableX_large.js:
```diff
@@ -0,0 +1,183 @@
+  sendTxAndGetReturnValue
+} = require("./utilities.js")
+
+const feeDenominator = 1000 // fee is (1 / feeDenominator)
+
+function feeSubtracted(x) {
+  return Math.floor(x * (feeDenominator - 1) / feeDenominator)
+}
+
+function feeAdded(x) {
+  return Math.floor(x * (feeDenominator) / (feeDenominator - 1))
+}
```
Maybe these functions can be required from a more general place.

bh2smith replied:
Yes, I can move them into utilities if you like.

bh2smith replied:
Wait.. this Branch is never going to be merged....

bh2smith commented:
> If we settle 40 orders , it is 15000*40 = 1/2 million more gas. Okay, it is not a lot, but probably significant. :)

So, based on our new max of ~25, this would be 
```
15_000 * 25 = 375_000
```
Which might be ok now that our 25 orders is shooting for about 4 million gas (i.e. a low estimate of half a block). We choose 4M because it is difficult/expensive to use an entire block.

bh2smith committed:
Merge branch 'master' into disregarded_utility

bh2smith committed:
Functional unit testing with only Utility (i.e. commented disregarded utility)

bh2smith committed:
Merge branch 'disregarded_utility' of github.com:gnosis/dex-contracts into disregarded_utility

bh2smith committed:
Fully functional U - dU with potentially missing unit tests.

bh2smith committed:
remove undefiend event emmision

bh2smith committed:
Update contract to use utility as objective function

bh2smith committed:
- Can't submit a solution with zero utility.
- Bunch of changes (in preparation for disregatded utility calculation)

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

# Conflicts:
#	contracts/StablecoinConverter.sol
#	test/stablecoin_converter.js

bh2smith committed:
no need for min int.

bh2smith committed:
Merge branch 'master' into utility_objective

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

bh2smith committed:
Use SafeMath and u256 in appropriate places

bh2smith committed:
small typo

bh2smith committed:
function can be restricted to view

bh2smith committed:
update revert message.

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

bh2smith committed:
use u256 before return

bh2smith committed:
No reason to have prices at 10.

bh2smith committed:
fix one test.

bh2smith committed:
fix one test.

bh2smith committed:
must verify limit price before evaluateUtility!

bh2smith committed:
must verify limit price before evaluateUtility!

bh2smith committed:
Merge branch 'disregarded_utility' into gastimate_utility

fleupold commented on test/stableX_large.js:
```diff
@@ -0,0 +1,183 @@
+  })
+
+  describe("submit single solutions", () => {
+    it("filling 10 orders", async () => {
```
Would we be able to reduce a lot of repetition here if we created a function fill_n_orders and called it with 10, 20 & 40 in each case?

bh2smith replied:
Already working on it

bh2smith committed:
update gas estimations with current U - dU

bh2smith committed:
Merge branch 'master' into utility_objective

bh2smith committed:
Introduce scaled utility

bh2smith committed:
Merge branch 'utility_objective' of github.com:gnosis/dex-contracts into utility_objective

bh2smith committed:
update utility evaluation in JS test file

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

bh2smith committed:
Merge branch 'disregarded_utility' into gastimate_utility

bh2smith committed:
remove commented line

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

bh2smith committed:
update comments in disregarded utility calculation.

bh2smith committed:
Merge branch 'disregarded_utility' into gastimate_utility

bh2smith committed:
Max touched orders (#232)

In order to prevent an unfair advantage to first solution submitter, we must restrict the number of touched orders to something that only uses no more than 4 Million gas (i.e. approx 25 orders). This is based on the gas estimations made in #231

bh2smith committed:
Update comment on utility evaluation and implement suggestion to pass fee rather than scale.

bh2smith committed:
include todo

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

# Conflicts:
#	contracts/StablecoinConverter.sol

bh2smith committed:
Merge branch 'disregarded_utility' of github.com:gnosis/dex-contracts into disregarded_utility

bh2smith committed:
update comment

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

bh2smith committed:
SafeMath and cast after division!

bh2smith committed:
update misleading comment

bh2smith committed:
easier objective evaluation

bh2smith committed:
Merge branch 'master' into utility_objective

bh2smith committed:
Merge branch 'master' into utility_objective

# Conflicts:
#	test/stablecoin_converter.js

bh2smith committed:
update unit tests for utility funciton (no need to capture the fees with volume vector).

bh2smith committed:
Merge branch 'utility_objective' into disregarded_utility

# Conflicts:
#	contracts/StablecoinConverter.sol
#	test/stablecoin_converter.js

bh2smith committed:
Merge branch 'disregarded_utility' into gastimate_utility

bh2smith committed:
Update utility to account for error.

bh2smith committed:
typo!

bh2smith committed:
forgot to take the floor

bh2smith committed:
Merge branch 'utility_error' into gastimate_utility

# Conflicts:
#	contracts/StablecoinConverter.sol
#	test/stablecoin_converter.js

bh2smith committed:
Update gas estimation to include utility error

bh2smith commented:
New estimates based on #266 and discussion in #257 


## Cost for  Submitting a Competing Solution

20 GWei with ETH @ ~183 USD

| num Orders   | Gas Used (Million) | USD Cost |
| ------------- |:-------------:| -----:|
| 10 | 2.0   | $7.24 |
| 20 | 3.8 |  $13.76 |
| 24 | 4.6 |  $16.65 |
| 30 | 5.6 |  $20.27 |
| 40 | gas limit exceeded | XXX |

bh2smith committed:
typo discovered as result of weird gas estimation.

bh2smith commented:
Now updated to include estimations for utility calculation with error term included (not much different)

bh2smith commented:
Closing since all related issues are merged (cf. #257 and #266).
